### ISSUE TITLE
Discussion: Embedded Vector Art and Procedural Content Generation

### ISSUE THREAD

**Issue by prndev** [Free-Cities/Free-Cities#889]
This is the continuation of the discussion sparked in #843.

@freecitiesdev wrote:

> the different boob sizes (...) are just resizes of a single base image. Rather than having seven .svgs  (...), would it be possible to have a single .svg that the script scales into 7+ options? That would save some space and clutter, or alternatively provide more options. Furthermore, it would make contributing items much easier as it would only be necessary to create a single, average size, to be scripted bigger and smaller.

As for the boobs: They actually are not exact resizes. Right now the larger boobs are drawn more "saggy" - as you would probably expect. However, we can totally drop these slight variations for the sake of procedural content generation. I expect, I would need to learn how svg transforms work, then I can dynamically generate an affine transformation matrix for automatic boob scaling.

That being said, I could use some help: It would be very kind if someone could look into `src/art/artWidgets.tw` (and the therein mentioned references to `src/art/vector/*`) in the [branch](prndev/Free-Cities/tree/embedded_vector_art_experiments) and port the case-switches (for selection of collars and alike) until it is mature enough to be merged into upstream. Then I could focus on the automatic scaling in the meantime. I cannot tell how much time I can devote for this.

_"Automatic boob scaling" - that's a term I never expected to write one day._

**Comment by prndev** [comment]
@lowercase-donkey wrote:
> If this is possible, tattoo and brand overlays start to become sane possibilities...

A proof of concept for this is now included in prndev/Free-Cities@14327ca (abdominal tattoo "rude words" only). :)

**Comment by prndev** [comment]
@freecitiesdev wrote:

> I'm not really seeing a downside

Please keep in mind that I had to tweak SugarCube so it supports inline-SVG. I cannot stress enough that **modifying the framework** always is a terrible idea. The modification can be done automatically for the next idunno updates of SugarCube, but in case SugarCube is drastically changed, I expect this automation to **fall apart terribly**.

**Comment by FCGudder** [comment]
> Please keep in mind that I had to tweak SugarCube so it supports inline-SVG.

Okay, I wanted to stay out of this conversation, but I have to ask ... why?!? SugarCube supports inline SVG just fine already. Here's a quick excerpt from the SVG filtering system over on pregmod:

```
:: SVG filters [widget]
<<widget "SVGFilters">>
<html>
<svg class="defs-only" style="width: 0; height: 0; position: absolute;">
	<filter id="skin-black" color-interpolation-filters="sRGB">
		<feColorMatrix type="matrix" values="0.15 0 0 0 0 0 0.1 0 0 0 0 0 0.05 0 0 0 0 0 1 0"/>
	</filter>
	<filter id="skin-bronzed" color-interpolation-filters="sRGB">
		<feColorMatrix type="matrix" values="0.8 0 0.1 0 0 0 0.6 0 0 0 0 0 0.3 0 0 0 0 0 1 0"/>
	</filter>
	<filter id="skin-brown" color-interpolation-filters="sRGB">
		<feColorMatrix type="matrix" values="0.8 0 0 0 0 0 0.5 0 0 0 0 0 0.15 0 0 0 0 0 1 0"/>
	</filter>
	<filter id="skin-dark" color-interpolation-filters="sRGB">
		<feColorMatrix type="matrix" values="0.5 0 0 0 0 0 0.2 0 0 0 0 0 0.1 0 0 0 0 0 1 0 "/>
	</filter>
(...)
	<filter id="hair-strawberry-blonde" color-interpolation-filters="sRGB">
		<feColorMatrix type="matrix" values="1.2 0 0 0 0.2 0 1.2 0 0 0 0 0 0.7 0 0 0 0 0 1 0"/>
	</filter>
	<filter id="hair-violet" color-interpolation-filters="sRGB">
		<feColorMatrix type="matrix" values="2 -1 0 0 0 0 2 -1 0 0 -1 0 2 0 0 0 0 0 1 0"/>
	</filter>
	<filter id="hair-white" color-interpolation-filters="sRGB">
		<feColorMatrix type="matrix" values="0.6 0 0 0 0.5 0 0.6 0 0 0.5 0 0 0.6 0 0.5 0 0 0 1 0"/>
	</filter>
	<filter id="hair-white-blonde" color-interpolation-filters="sRGB">
		<feColorMatrix type="matrix" values="0.7 0 0 0 0.5 0 0.7 0 0 0.5 0 0 0.5 0 0.4 0 0 0 1 0"/>
	</filter>
</svg>
</html>
<</widget>>
```
That's all. It works. Please don't try to "fix" what isn't broken.

For reference, pregmod's dynamic skin and hair colour implementation can be found at gitgud.io/pregmodfan/fc-pregmod/commit/f32b69ac639326549ec5af0d4b58f4298929d0b0 and a test game (with lots of other unrelated tweaks) is at fcgudder.gitlab.io.

**Comment by prndev** [comment]
Oh, this whole thing already exists. Too bad I didn't notice you sooner.

I read the twine documentation and on  https://twinery.org/wiki/twine2:add_html it tells you just to put HTML directly into the story. The SugarCube documentation at https://www.motoslave.net/sugarcube/2/docs/ mentions a script macro for raw JavaScript, but not for HTML. Therefore I assumed, it was not possible by default. It is nice to know this feature already exists although makes my efforts look stupid.

**Comment by lowercase-donkey** [comment]
@prndev It might suck to feel like work was wasted, but it's pretty amazing not having to worry about your work breaking with every Twine update.

**Comment by freecitiesdev** [comment]
If you're still willing to implement this, I strongly support it. Flexible colors would be great for more skin and hair tones, and hopefully for eyes if implemented. Proper .svg scaling would allow tattoos and boob shapes. The list goes on; I'm sure there's other great stuff.

**Comment by prndev** [comment]
Implementing boob scaling was less cumbersome than I expected. The fun part is in `src/art/vector/Boob.tw`. I had to make some changes to the artwork, tough. Please feel free to checkout my [experimental branch](prndev/Free-Cities/tree/boob_scaling_experiments) and tell me what you think. It works during my play-tests so far. If we find no problems within the next days, I am willing to merge the experimental branch with the current master, #843 and #884 and then create a pull request.

**Comment by freecitiesdev** [comment]
Closing this due to implementation. Open a new issue for discussion after #923.
